influxdb3_client/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+    /// The base URL for making requests to a running InfluxDB 3.0 server
+    base_url: Url,
+    /// The `Bearer` token to use for authenticating on each request to the server
+    auth_header: Option<Secret<String>>,
+    /// A [`reqwest::Client`] for handling HTTP requests
+    http_client: reqwest::Client,
+}
+
+impl Client {
+    /// Create a new [`Client`]
+    ///
+    /// A [`reqwest::Client`] is passed in order to allow for re-use of an existing
+    /// `reqwest::Client`.
+    pub fn new<U: IntoUrl>(base_url: U, http_client: reqwest::Client) -> Result<Self> {
+        Ok(Self {
+            base_url: base_url.into_url().map_err(Error::BaseUrl)?,
+            auth_header: None,
+            http_client,
+        })
+    }
+
+    /// Set the `Bearer` token that will be sent with each request to the server
+    pub fn with_auth_header<S: Into<String>>(mut self, auth_header: S) -> Self {
+        self.auth_header = Some(Secret::new(auth_header.into()));
+        self
+    }
+
+    /// Compose a request to the `/api/v3/write_lp` API
+    ///
+    /// # Example
+    /// ```no_run
+    /// # use influxdb3_client::Client;
+    /// # use influxdb3_client::Precision;
+    /// # #[tokio::main]
+    /// # async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
+    /// let client = Client::new("localhost:8181", reqwest::Client::new())?;
+    /// client
+    ///     .api_v3_write_lp("db_name")
+    ///     .precision(Precision::Milli)
+    ///     .accept_partial(true)
+    ///     .body("cpu,host=s1 usage=0.5")
+    ///     .send()
+    ///     .await
+    ///     .expect("send write_lp request");
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn api_v3_write_lp<S: Into<String>>(&self, db: S) -> WriteRequestBuilder<'_, NoBody> {
+        WriteRequestBuilder {
+            client: self,
+            db: db.into(),
+            precision: None,
+            accept_partial: None,
+            body: NoBody,
+        }
+    }
+
+    /// Compose a request to the `/api/v3/query_sql` API
+    ///
+    /// # Example
+    /// ```no_run
+    /// # use influxdb3_client::Client;
+    /// # #[tokio::main]
+    /// # async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
+    /// let client = Client::new("localhost:8181", reqwest::Client::new())?;
+    /// let response_bytes = client
+    ///     .api_v3_query_sql("db_name", "SELECT * FROM foo")
+    ///     .send()
+    ///     .await
+    ///     .expect("send query_sql request");
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn api_v3_query_sql<D: Into<String>, Q: Into<String>>(
+        &self,
+        db: D,
+        query: Q,
+    ) -> QueryRequestBuilder<'_> {
+        QueryRequestBuilder {
+            client: self,
+            db: db.into(),
+            query: query.into(),
+            format: None,
+        }
+    }
+}
+
+/// The body of the request to the `/api/v3/write_lp` API
+// TODO - this should re-use the type defined in the server code.
```

Comment:
I think it's outside the scope of this PR, but I think shared types between the client and server should be their own crate. I have a feeling that this crate in particular might be something we release independent of Edge for consumers to use.

Reply:
Issue created: #24672